client/client-lib/src/modules.rs

```diff
@@ -0,0 +1,260 @@
+use std::{collections::BTreeMap, io, sync::Arc};
+
+use impl_tools::autoimpl;
+
+use bitcoin::Amount;
+use thiserror::Error;
+
+/// Object-safe trait for things that can encode themselves
+///
+/// Like `rust-bitcoin`'s `consensus_encode`, but without generics,
+/// so can be used in `dyn` objects.
+pub trait Encodable {
+    fn consensus_encode(&self, writer: &mut dyn io::Write) -> Result<usize, std::io::Error>;
+}
+
+/// Something that can be an [`Input`] in a [`Transaction`]
+///
+/// General purpose code should use [`Input`] instead
+pub trait IInput: Encodable {}
+
+/// Something that can be an [`Output`] in a [`Transaction`]
+///
+/// General purpose code should use [`Output`] instead
+pub trait IOutput: Encodable {}
+
+/// An owned, immutable input to a [`Transaction`]
+#[autoimpl(Deref using self.0)]
+pub struct Input(Box<dyn IInput + Send + Sync + 'static>);
+
+/// An owned, immutable output of a [`Transaction`]
+#[autoimpl(Deref using self.0)]
+pub struct Output(Box<dyn IOutput + Send + Sync + 'static>);
+
+/// A spendable output - tracked and persisted by the client
+///
+/// Created by generating transaction [`Output`], spendable
+/// by converting to [`Input`].
+pub trait ISpendableOutput {
```

Comment:
This is an important addition. Basically a job of a client wallet is tracking these, and then using them to build transactions, which produce more of these, and so on.